examples/quickstart-monai/data.py

```diff
@@ -0,0 +1,153 @@
+import os
+import tarfile
+from urllib import request
+
+import numpy as np
+from monai.data import DataLoader, Dataset
+from monai.transforms import (
+    Compose,
+    EnsureChannelFirst,
+    LoadImage,
+    RandFlip,
+    RandRotate,
+    RandZoom,
+    ScaleIntensity,
+    ToTensor,
+)
+from torch.utils.data import Subset
+
+
+def _partition(dataset, num_shards, index):
+    total_size = len(dataset)
+    shard_size = total_size // num_shards
+
+    # Calculate start and end indices for the shard
+    start_idx = index * shard_size
+    if index == num_shards - 1:
+        # Last shard takes the remainder
+        end_idx = total_size
+    else:
+        end_idx = start_idx + shard_size
+
+    # Create a subset for the shard
+    shard = Subset(dataset, range(start_idx, end_idx))
+    return shard
+
+
+def load_data(num_shards, index):
+    image_file_list, image_label_list, num_total, num_class = _download_data()
+    trainX, trainY, valX, valY, testX, testY = _split_data(
+        image_file_list, image_label_list, num_total
+    )
+    train_transforms, val_transforms = _get_transforms()
+
+    train_ds = _partition(
+        MedNISTDataset(trainX, trainY, train_transforms), num_shards, index
+    )
+    train_loader = DataLoader(train_ds, batch_size=300, shuffle=True)
+
+    val_ds = _partition(MedNISTDataset(valX, valY, val_transforms), num_shards, index)
+    val_loader = DataLoader(val_ds, batch_size=300)
+
+    test_ds = _partition(
+        MedNISTDataset(testX, testY, val_transforms), num_shards, index
+    )
```

Comment:
```suggestion


def _partition(files_list, labels_list, num_shards, index):
    total_size = len(files_list)
    assert total_size == len(labels_list), f"List of datapoints and labels must be of the same length"
    shard_size = total_size // num_shards

    # Calculate start and end indices for the shard
    start_idx = index * shard_size
    if index == num_shards - 1:
        # Last shard takes the remainder
        end_idx = total_size
    else:
        end_idx = start_idx + shard_size

    # Create a subset for the shard
    files = files_list[start_idx: end_idx]
    labels = labels_list[start_idx: end_idx]
    return files, labels


def load_data(num_shards, index):
    image_file_list, image_label_list, num_total, num_class = _download_data()

    # get partition given index
    files_list, labels_list = _partition(image_file_list, image_label_list, num_shards, index)


    trainX, trainY, valX, valY, testX, testY = _split_data(
        files_list, labels_list, len(files_list)
    )
    train_transforms, val_transforms = _get_transforms()

    train_ds = MedNISTDataset(trainX, trainY, train_transforms)
    train_loader = DataLoader(train_ds, batch_size=300, shuffle=True)

    val_ds = MedNISTDataset(valX, valY, val_transforms)
    val_loader = DataLoader(val_ds, batch_size=300)

    test_ds = MedNISTDataset(testX, testY, val_transforms)
    test_loader = DataLoader(test_ds, batch_size=300)

    return train_loader, val_loader, test_loader, num_class
```

Reply:
While trying to rewrite how datasets are created (attempting to fix the segfault issue) I came up with this slightly simplified version of `load_data`. In short: partition first, then create datasets (as opposed of creating full datasets and then extract subsets)